File: courses/codes/cclv1-jr-class/median_finder.py
def insertion_sort(A):
    for i in range(1, len(A)):
        temp = A[i]
        j = i
        while j > 0 and temp < A[j - 1]:
            A[j] = A[j - 1]
            j -= 1
        A[j] = temp
    return A

def selection_sort(A):
    for i in range(len(A) - 1, -1, -1):
        max_val, max_loc = A[0], 0
        for j in range(1, i):
            if max_val < A[j]:
                max_val, max_loc = A[j], j
        A[i], A[max_loc] = A[max_loc], A[i]
    return A

def bubble_sort(A):
    for i in range(len(A) - 1, -1, -1):
        for j in range(i):
            if A[j] > A[j + 1]:
                A[j], A[j + 1] = A[j + 1], A[j]
    return A

n = int(input())
nums = []
while n > 0:
    nums.append(n)
    nums = insertion_sort(nums)     # can be any of the three sorting algorithms
    if len(nums) % 2 == 1:
        median = nums[len(nums) // 2]
    else:
        median = (nums[len(nums) // 2 - 1] + nums[len(nums) // 2]) / 2
        if median == int(median):
            median = int(median)
    print("The median so far is " + str(median) + ".")
    n = int(input())
print("End of program.")

File: courses/codes/cclv1-jr-class/MedianFinder.java
import java.util.Scanner;
import java.util.Arrays;

public class MedianFinder {

	public static int[] insertionSort(int[] A) {
		for (int i = 1; i < A.length; i++) {
			int temp = A[i];
			int j = i;
			while (j > 0 && temp < A[j - 1]) {
				A[j] = A[j - 1];
				j--;
			}
			A[j] = temp;
		}
		return A;
	}
	
	public static int[] selectionSort(int[] A) {
		for (int i = A.length - 1; i > 0; i--) {
			int max = A[0];
			int maxLoc = 0;
			for (int j = 1; j <= i; j++) {
				if (max < A[j]) {
					max = A[j];
					maxLoc = j;
				}
			}
			int temp = A[i];
			A[i] = A[maxLoc];
			A[maxLoc] = temp;
		}
		return A;
	}
	
	public static int[] bubbleSort(int[] A) {
		for (int i = A.length - 1; i >= 0; i--) {
			for (int j = 0; j < i; j++) {
				if (A[j] > A[j + 1]) {
					int temp = A[j];
					A[j] = A[j + 1];
					A[j + 1] = temp;
				}
			}
		}
		return A;
	}
	
	public static void main(String[] args) {
		Scanner sc = new Scanner(System.in);
		int n = sc.nextInt();
		int[] nums = new int[1];
		while (n > 0) {
			int lastSpot = nums.length - 1;
			nums[lastSpot] = n;
			nums = insertionSort(nums);		// feel free to use any sorting algorithm
			
			if (nums.length % 2 == 1) {
				System.out.println("The median so far is " + nums[nums.length / 2] + ".");
			} else {
				double num1 = (double) nums[nums.length / 2 - 1];
				double num2 = (double) nums[nums.length / 2];
				double med = (num1 + num2) / 2;
				int intMed = (int) med;
				if (med == intMed) {
					System.out.println("The median so far is " + intMed + ".");
				} else {
					System.out.println("The median so far is " + med + ".");
				}
			}
			n = sc.nextInt();
			int[] tempArr = new int[nums.length + 1];
			System.arraycopy(nums, 0, tempArr, 0, nums.length);
			nums = new int[nums.length + 1];
			System.arraycopy(tempArr, 0, nums, 0, tempArr.length);
		}
		System.out.println("End of program.");
	}

}
